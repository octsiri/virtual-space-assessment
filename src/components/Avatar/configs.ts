export enum AvatarSize {
  'xs',
  'sm',
  'md',
  'lg',
}

export type AvatarSizeTypes = keyof typeof AvatarSize;

export const avatarSizeMapper = {
  xs: {
    size: 24,
    fontSize: 14,
    borderWidth: 2,
    borderRadius: 6,
    marginLeft: -8,
  },
  sm: {
    size: 32,
    fontSize: 18,
    borderWidth: 2,
    borderRadius: 6,
    marginLeft: -10,
  },
  md: {
    size: 48,
    fontSize: 28,
    borderWidth: 3,
    borderRadius: 6,
    marginLeft: -12,
  },
  lg: {
    size: 64,
    fontSize: 40,
    borderWidth: 3.5,
    borderRadius: 6,
    marginLeft: -14,
  },
};
