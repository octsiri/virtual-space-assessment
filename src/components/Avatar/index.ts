export {default as Avatar} from './Avatar';
export {default as AvatarsGroup} from './AvatarsGroup';
