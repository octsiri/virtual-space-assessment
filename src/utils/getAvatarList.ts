export const getAvatarList = () => {
  return [
    {
      id: '1',
      username: 'Julius Cesaren',
      imageSource:
        'https://t4.ftcdn.net/jpg/03/64/21/11/240_F_364211147_1qgLVxv1Tcq0Ohz3FawUfrtONzz8nq3e.jpg',
    },
    {
      id: '2',
      username: 'Tomi Budi',
      imageSource: '',
    },
    {
      id: '3',
      username: 'Angelina Joli',
      imageSource:
        'https://st.depositphotos.com/1694341/1303/i/600/depositphotos_13037294-stock-photo-angelina-jolie.jpg',
    },
    {
      id: '4',
      username: 'Tje Jody',
      imageSource: 'https://thumbs.dreamstime.com/b/jodi-lyn-o-keefe-26041224.jpg',
    },
    {
      id: '5',
      username: 'Dominggus Octovianus',
      imageSource: '',
    },
    {
      id: '6',
      username: 'Neymar Jr',
      imageSource:
        'https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcSZHR_H3Tm3nQNIhpIniTDKGE2jqyV-L4bWXA&usqp=CAU',
    },
  ];
};
