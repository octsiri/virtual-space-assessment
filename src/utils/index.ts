export {default as convertValueToUnit} from './convertValueToUnit';
export {default as getInitialCapitalName} from './getInitialCapitalName';
