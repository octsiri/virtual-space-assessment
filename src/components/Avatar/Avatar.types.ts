import * as React from 'react';

import {ColorsType} from '../../foundations';
import {AvatarSizeTypes} from './configs';

export interface Props {
  hidden?: boolean;
  src?: string;
  username?: string;
  size: AvatarSizeTypes;
  noTrim?: boolean;
  bordered?: boolean;
  withPadding?: boolean;
  defaultColor?: ColorsType;
  borderColor?: ColorsType;
  defaultBackground?: ColorsType;
  style?: React.CSSProperties;
  onClick?: () => void;
}

export interface AvatarImageContainerProps {
  name?: string;
  withPadding?: boolean;
  size?: AvatarSizeTypes;
  src?: string;
  alt?: string;
  defaultBackground?: string;
  clickable?: boolean;
}
