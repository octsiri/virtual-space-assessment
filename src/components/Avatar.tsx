import * as React from 'react';
import styled, {css} from 'styled-components';
import {colors, ColorsType, Image, Text, View} from '../foundations';
import {avatarSizeMapper, type AvatarSizeTypes, getInitialCapitalName} from '../utils';

interface Props {
  hidden?: boolean;
  src?: string;
  name?: string;
  size: AvatarSizeTypes;
  noTrim?: boolean;
  bordered?: boolean;
  withPadding?: boolean;
  defaultColor?: ColorsType;
  borderColor?: ColorsType;
  defaultBackground?: ColorsType;
  style?: React.CSSProperties;
  onClick?: () => void;
}

interface AvatarImageContainerProps {
  name?: string;
  withPadding?: boolean;
  size?: AvatarSizeTypes;
  src?: string;
  alt?: string;
  defaultBackground?: string;
}

const AvatarWrapper = styled.div<Props>`
  display: flex;
  align-items: flex-end;
  justify-content: flex-end;
  ${(props) =>
    props.bordered
      ? css`
          border-color: ${props.borderColor ? colors[props.borderColor] : 'white'};
          border-style: solid;
          border-width: ${avatarSizeMapper[props.size].borderWidth}px;
        `
      : ''};
  width: ${(props) => avatarSizeMapper[props.size].size}px;
  height: ${(props) => avatarSizeMapper[props.size].size}px;
  font-weight: 500;
  border-radius: 50%;
  background: ${(props) => (props.defaultBackground ? props.defaultBackground : colors.smokeLight)};
  position: relative;
  text-align: center;
`;

const AvatarImageContainer = styled.div<AvatarImageContainerProps>`
  display: flex;
  align-items: center;
  justify-content: center;
  width: 100%;
  height: 100%;
  overflow: hidden;
  background: ${(props) => (props.defaultBackground ? props.defaultBackground : colors.smokeLight)};
  color: ${(props) => (props.color ? props.color : colors.gray60)};
  border-radius: 50%;
  &:hover {
    cursor: pointer;
    opacity: 0.9;
  }
`;

const Avatar = (props: Props) => {
  const {
    hidden,
    src,
    name,
    size,
    noTrim,
    bordered,
    borderColor = 'white',
    defaultColor = 'cyanDarkest',
    defaultBackground = 'blueLightest',
    onClick,
    style,
  } = props;
  const background = colors[defaultBackground];
  const color = colors[defaultColor];

  const avatar = React.useMemo(() => {
    if (src && src.length) {
      return (
        <Image
          background={`url(${src})`}
          width={`${avatarSizeMapper[size].size}px`}
          objectFit='cover'
        />
      );
    }
    return (
      <View flexDirection='row' flexAlignItems='center' flexJustifyContent='center'>
        {noTrim ? (
          <Text size={size} color={defaultColor} textAlign='justify' style={{marginTop: -3}}>
            +
          </Text>
        ) : (
          <div />
        )}
        <Text size={size} color={defaultColor} textAlign='justify'>
          {noTrim ? name : getInitialCapitalName(name, 2)}
        </Text>
      </View>
    );
  }, [src, name, size, noTrim]);

  return !hidden ? (
    <AvatarWrapper
      bordered={bordered}
      size={size}
      defaultBackground={defaultBackground}
      borderColor={borderColor}
      style={style}
      onClick={onClick}
    >
      <AvatarImageContainer defaultBackground={background} color={color} size={size}>
        {avatar}
      </AvatarImageContainer>
    </AvatarWrapper>
  ) : (
    <div />
  );
};

export default Avatar;
