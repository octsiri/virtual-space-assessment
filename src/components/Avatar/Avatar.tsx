import * as React from 'react';
import styled, {css} from 'styled-components';

import {colors, Image, Text, View} from '../../foundations';
import {getInitialCapitalName} from '../../utils';
import {Props, AvatarImageContainerProps} from './Avatar.types';
import {avatarSizeMapper} from './configs';

const AvatarWrapper = styled.div<Props>`
  display: flex;
  align-items: flex-end;
  justify-content: flex-end;
  border-radius: 50%;
  ${(props) =>
    props.bordered
      ? css`
          border-color: ${props.borderColor ? colors[props.borderColor] : 'white'};
          border-style: solid;
          border-width: ${avatarSizeMapper[props.size || 'md'].borderWidth}px;
        `
      : ''};
  width: ${(props) => avatarSizeMapper[props.size || 'md'].size}px;
  height: ${(props) => avatarSizeMapper[props.size || 'md'].size}px;
  background-color: ${(props) =>
    props.defaultBackground ? colors[props.defaultBackground] : colors.gray70};
  position: relative;
  text-align: center;
`;

const AvatarImageContainer = styled.div<AvatarImageContainerProps>`
  display: flex;
  align-items: center;
  justify-content: center;
  width: 100%;
  height: 100%;
  overflow: hidden;
  background: ${(props) => (props.defaultBackground ? props.defaultBackground : colors.smokeLight)};
  color: ${(props) => (props.color ? props.color : colors.gray70)};
  border-radius: 50%;
  &:hover {
    cursor: ${(props) => (props.clickable ? 'pointer' : undefined)};
    opacity: 0.9;
  }
`;

const Avatar = (props: Props) => {
  const {
    hidden,
    src,
    username,
    noTrim,
    bordered,
    size = 'md',
    borderColor = 'gray20',
    defaultColor = 'gray70',
    defaultBackground = 'cyanLightest',
    onClick,
    style,
  } = props;
  const background = colors[defaultBackground];
  const color = colors[defaultColor];

  // NOTE: This is for adjust the text size for number more than a digit
  const customSize = React.useMemo(() => {
    if (noTrim && username && username.length > 1) {
      switch (size) {
        case 'xs':
        case 'sm':
          return 'xs';
        case 'md':
          return 'sm';
        default:
          return 'md';
      }
    }
    return size;
  }, [size, noTrim, username]);

  const avatar = React.useMemo(() => {
    if (src && src.length) {
      return (
        <Image
          alt='avatar-image'
          background={`url(${src})`}
          width={`${avatarSizeMapper[size].size}px`}
          objectFit='cover'
        />
      );
    }
    return (
      <View flexDirection='row' flexAlignItems='center' flexJustifyContent='center'>
        {noTrim ? (
          <Text size={customSize} color={defaultColor} textAlign='justify' style={{marginTop: -3}}>
            +
          </Text>
        ) : (
          <div />
        )}
        <Text size={customSize} color={defaultColor} textAlign='justify'>
          {noTrim ? username : getInitialCapitalName(username, 2)}
        </Text>
      </View>
    );
  }, [src, username, size, noTrim]);

  return !hidden ? (
    <AvatarWrapper
      bordered={bordered}
      size={size}
      defaultBackground={defaultBackground}
      borderColor={borderColor}
      style={style}
      onClick={onClick}
    >
      <AvatarImageContainer
        defaultBackground={background}
        color={color}
        size={size}
        clickable={Boolean(onClick)}
      >
        {avatar}
      </AvatarImageContainer>
    </AvatarWrapper>
  ) : (
    <div />
  );
};

export default React.memo<Props>(Avatar);
