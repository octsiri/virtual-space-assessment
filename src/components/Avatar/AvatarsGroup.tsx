import * as React from 'react';
import styled from 'styled-components';
import {AvatarSizeTypes, avatarSizeMapper} from './configs';
import Avatar from './Avatar';

interface AvatarDataType {
  id: string;
  username: string | undefined;
  imageSource: string | undefined;
}

interface Props {
  avatars?: AvatarDataType[];
  maxLength?: number;
  size?: AvatarSizeTypes;
}

const Wrapper = styled.div`
  display: flex;
  flex-wrap: wrap;
  flex-direction: row;
  justify-content: center;
  align-items: center;
`;

const AvatarsGroup = (props: Props) => {
  const {avatars = [], maxLength = 1, size = 'md'} = props;
  const avatarRemains = avatars ? avatars.length - maxLength : -1;

  const avatarGroupContents = React.useMemo(() => {
    if (avatars && avatars.length) {
      return avatars.map((avatar, idx) => (
        <Avatar
          key={'avatar-' + idx + avatar.id}
          src={avatar.imageSource}
          username={avatar.username}
          size={size}
          bordered
          borderColor='white'
          hidden={idx >= (maxLength as number)}
          style={{marginLeft: avatarSizeMapper[size].marginLeft}}
        />
      ));
    }
  }, [avatars, size, maxLength]);

  if (avatars && avatars.length) {
    return (
      <Wrapper>
        {avatarGroupContents}
        <Avatar
          noTrim
          username={`${avatarRemains}`}
          size={size}
          bordered
          borderColor='white'
          hidden={avatarRemains <= 0}
          style={{marginLeft: avatarSizeMapper[size].marginLeft}}
        />
      </Wrapper>
    );
  }
  return <Avatar />;
};

export default React.memo<Props>(AvatarsGroup);
