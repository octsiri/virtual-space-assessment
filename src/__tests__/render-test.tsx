/* eslint-env jest */
import React from 'react';
import renderer from 'react-test-renderer';

import {Avatar, AvatarsGroup} from '../components/Avatar';
import {getAvatarList} from '../utils/getAvatarList';

describe('Snapshot Avatar Component', () => {
  test('Render avatar with default props correctly', () => {
    const tree = renderer.create(<Avatar />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('Render avatar with custom props correctly', () => {
    const tree = renderer
      .create(
        <Avatar
          src='https://t4.ftcdn.net/jpg/03/64/21/11/240_F_364211147_1qgLVxv1Tcq0Ohz3FawUfrtONzz8nq3e.jpg'
          size='md'
          onClick={() => {
            console.log('click');
          }}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('Render avatar with initial username only correctly', () => {
    const tree = renderer
      .create(
        <Avatar
          username='Tony Shark'
          onClick={() => {
            console.log('click');
          }}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('Render avatar with people remains counts and border props correctly', () => {
    const tree = renderer
      .create(
        <Avatar
          noTrim
          username='7'
          bordered
          borderColor='gray20'
          defaultBackground='blue'
          defaultColor='cyan'
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('Render avatar with hidden props true correctly', () => {
    const tree = renderer
      .create(
        <Avatar
          hidden
          src='https://t4.ftcdn.net/jpg/03/64/21/11/240_F_364211147_1qgLVxv1Tcq0Ohz3FawUfrtONzz8nq3e.jpg'
          username='Invisble Man'
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('Snapshot AvatarsGroup Component', () => {
  test('Render avatars group with default props correctly', () => {
    const tree = renderer.create(<AvatarsGroup />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('Render avatars group with custom props correctly', () => {
    const tree = renderer
      .create(<AvatarsGroup avatars={getAvatarList()} size='lg' maxLength={3} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
