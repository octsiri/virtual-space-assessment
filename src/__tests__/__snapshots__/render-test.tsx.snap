// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot Avatar Component Render avatar with custom props correctly 1`] = `
<div
  className="sc-hLBbgP bgpjRe"
  onClick={[Function]}
  size="md"
>
  <div
    className="sc-eDvSVe bFTEKl"
    color="#767676"
    size="md"
  >
    <div
      alt="avatar-image"
      className="sc-bcXHqe gLpgRM"
      height="100%"
      style={
        Object {
          "background": "url(https://t4.ftcdn.net/jpg/03/64/21/11/240_F_364211147_1qgLVxv1Tcq0Ohz3FawUfrtONzz8nq3e.jpg) center center / cover repeat",
        }
      }
      width="48px"
    />
  </div>
</div>
`;

exports[`Snapshot Avatar Component Render avatar with default props correctly 1`] = `
<div
  className="sc-hLBbgP bgpjRe"
  size="md"
>
  <div
    className="sc-eDvSVe fclrjC"
    color="#767676"
    size="md"
  >
    <div
      className="sc-dkrFOg cVibjD"
    >
      <div />
      <span
        className="sc-gswNZR dRPPka"
        color="gray70"
        size="md"
      >
        U
      </span>
    </div>
  </div>
</div>
`;

exports[`Snapshot Avatar Component Render avatar with hidden props true correctly 1`] = `<div />`;

exports[`Snapshot Avatar Component Render avatar with initial username only correctly 1`] = `
<div
  className="sc-hLBbgP bgpjRe"
  onClick={[Function]}
  size="md"
>
  <div
    className="sc-eDvSVe bFTEKl"
    color="#767676"
    size="md"
  >
    <div
      className="sc-dkrFOg cVibjD"
    >
      <div />
      <span
        className="sc-gswNZR dRPPka"
        color="gray70"
        size="md"
      >
        TS
      </span>
    </div>
  </div>
</div>
`;

exports[`Snapshot Avatar Component Render avatar with people remains counts and border props correctly 1`] = `
<div
  className="sc-hLBbgP cSAVTH"
  size="md"
>
  <div
    className="sc-eDvSVe viaQW"
    color="#04A4A4"
    size="md"
  >
    <div
      className="sc-dkrFOg cVibjD"
    >
      <span
        className="sc-gswNZR cZNUXP"
        color="cyan"
        size="md"
        style={
          Object {
            "marginTop": -3,
          }
        }
      >
        +
      </span>
      <span
        className="sc-gswNZR cZNUXP"
        color="cyan"
        size="md"
      >
        7
      </span>
    </div>
  </div>
</div>
`;

exports[`Snapshot AvatarsGroup Component Render avatars group with custom props correctly 1`] = `
<div
  className="sc-jSUZER bFvggE"
>
  <div
    className="sc-hLBbgP cWohFM"
    size="lg"
    style={
      Object {
        "marginLeft": -14,
      }
    }
  >
    <div
      className="sc-eDvSVe fclrjC"
      color="#767676"
      size="lg"
    >
      <div
        alt="avatar-image"
        className="sc-bcXHqe iRhMtq"
        height="100%"
        style={
          Object {
            "background": "url(https://t4.ftcdn.net/jpg/03/64/21/11/240_F_364211147_1qgLVxv1Tcq0Ohz3FawUfrtONzz8nq3e.jpg) center center / cover repeat",
          }
        }
        width="64px"
      />
    </div>
  </div>
  <div
    className="sc-hLBbgP cWohFM"
    size="lg"
    style={
      Object {
        "marginLeft": -14,
      }
    }
  >
    <div
      className="sc-eDvSVe fclrjC"
      color="#767676"
      size="lg"
    >
      <div
        className="sc-dkrFOg cVibjD"
      >
        <div />
        <span
          className="sc-gswNZR BFEHI"
          color="gray70"
          size="lg"
        >
          TB
        </span>
      </div>
    </div>
  </div>
  <div
    className="sc-hLBbgP cWohFM"
    size="lg"
    style={
      Object {
        "marginLeft": -14,
      }
    }
  >
    <div
      className="sc-eDvSVe fclrjC"
      color="#767676"
      size="lg"
    >
      <div
        alt="avatar-image"
        className="sc-bcXHqe iRhMtq"
        height="100%"
        style={
          Object {
            "background": "url(https://st.depositphotos.com/1694341/1303/i/600/depositphotos_13037294-stock-photo-angelina-jolie.jpg) center center / cover repeat",
          }
        }
        width="64px"
      />
    </div>
  </div>
  <div />
  <div />
  <div />
  <div
    className="sc-hLBbgP cWohFM"
    size="lg"
    style={
      Object {
        "marginLeft": -14,
      }
    }
  >
    <div
      className="sc-eDvSVe fclrjC"
      color="#767676"
      size="lg"
    >
      <div
        className="sc-dkrFOg cVibjD"
      >
        <span
          className="sc-gswNZR BFEHI"
          color="gray70"
          size="lg"
          style={
            Object {
              "marginTop": -3,
            }
          }
        >
          +
        </span>
        <span
          className="sc-gswNZR BFEHI"
          color="gray70"
          size="lg"
        >
          3
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`Snapshot AvatarsGroup Component Render avatars group with default props correctly 1`] = `<div />`;
